mlir/lib/Dialect/Func/TransformOps/FuncTransformOps.cpp

```diff
@@ -226,6 +227,98 @@ void transform::CastAndCallOp::getEffects(
+           << "function with name '" << getFunctionName() << "' not found";
+
+  int numArgs = funcOp.getNumArguments();
+  int numResults = funcOp.getNumResults();
+  // Check that the number of arguments and results matches the
+  // interchange sizes.
+  if (numArgs != (int)getArgsInterchange().size())
+    return emitSilenceableFailure(getLoc())
+           << "function with name '" << getFunctionName() << "' has " << numArgs
+           << " arguments, but " << getArgsInterchange().size()
+           << " args interchange were given";
+
+  if (numResults != (int)getResultsInterchange().size())
+    return emitSilenceableFailure(getLoc())
+           << "function with name '" << getFunctionName() << "' has "
+           << numResults << " results, but " << getResultsInterchange().size()
+           << " results interchange were given";
+
+  // Check that the args and results interchanges are unique.
+  SetVector<int> argsInterchange, resultsInterchange;
+  argsInterchange.insert_range(getArgsInterchange());
+  resultsInterchange.insert_range(getResultsInterchange());
+  if (argsInterchange.size() != getArgsInterchange().size())
+    return emitSilenceableFailure(getLoc())
+           << "args interchange must be unique";
+
+  if (resultsInterchange.size() != getResultsInterchange().size())
+    return emitSilenceableFailure(getLoc())
+           << "results interchange must be unique";
+
+  // Check that the args and results interchange indices are in bounds.
+  for (auto index : argsInterchange) {
+    if (index < 0 || index >= numArgs) {
+      return emitSilenceableFailure(getLoc())
+             << "args interchange index " << index
+             << " is out of bounds for function with name '"
+             << getFunctionName() << "' with " << numArgs << " arguments";
+    }
+  }
+  for (auto index : resultsInterchange) {
+    if (index < 0 || index >= numResults) {
+      return emitSilenceableFailure(getLoc())
+             << "results interchange index " << index
+             << " is out of bounds for function with name '"
+             << getFunctionName() << "' with " << numResults << " results";
+    }
+  }
+
+  func::replaceFuncWithNewOrder(funcOp, argsInterchange.getArrayRef(),
+                                resultsInterchange.getArrayRef());
+  if (getAdjustFuncCalls()) {
+    SmallVector<func::CallOp> callOps;
+    targetModuleOp.walk([&](func::CallOp callOp) {
+      if (callOp.getCallee() == getFunctionName().getRootReference().str())
+        callOps.push_back(callOp);
+    });
+
+    for (auto callOp : callOps)
+      func::replaceCallOpWithNewOrder(callOp, argsInterchange.getArrayRef(),
+                                      resultsInterchange.getArrayRef());
+  }
+
+  return DiagnosedSilenceableFailure::success();
+}
+
+void transform::ReplaceFuncSignatureOp::getEffects(
+    SmallVectorImpl<MemoryEffects::EffectInstance> &effects) {
+  transform::consumesHandle(getModuleMutable(), effects);
```

Comment:
Let's return a new handle so transforms can compose with each other. Otherwise the potentially top-level handle becomes unusable after this transformation.